### PR TITLE
Use tf.keras.metrics

### PR DESCRIPTION
**Pull request by tabergma**
**Proposed changes**:
- part of https://github.com/RasaHQ/research/issues/62

**Status (please check what you already did)**:
- [ ] added some tests for the functionality
- [ ] updated the documentation
- [ ] updated the changelog (please check [changelog](https://github.com/RasaHQ/rasa/tree/master/changelog) for instructions)
- [x] reformat files using `black` (please check [Readme](https://github.com/RasaHQ/rasa#code-style) for instructions)

### CODE REVIEW

File: rasa/nlu/classifiers/embedding_intent_classifier.py [hunk 1]
```diff
@@ -848,48 +848,14 @@ def load(
...
-
     def __init__(
         self,
         data_signature: Dict[Text, List[FeatureSignature]],
         label_data: RasaModelData,
         inverted_tag_dict: Dict[int, Text],
         config: Dict[Text, Any],
     ) -> None:
-        super(DIET, self).__init__(name="DIET")
+        super().__init__()
```
Comment (Ghostvv) [inline rasa/nlu/classifiers/embedding_intent_classifier.py]: let's keep the name: `name="DIET"`

File: rasa/nlu/classifiers/embedding_intent_classifier.py [hunk 2]
```diff
@@ -1024,23 +997,21 @@ def _prepare_layers(self) -> None:
...
             self._embed["logits"] = tf_layers.Embed(
                 self._num_tags, self.config[C2], "logits"
             )
             self._crf = tf_layers.CRF(self._num_tags, self.config[C2])
-            self.train_metrics["e_loss"] = tf.keras.metrics.Mean(name="e_loss")
-            self.train_metrics["e_f1"] = tf.keras.metrics.Mean(name="e_f1")
-            self.eval_metrics["val_e_loss"] = tf.keras.metrics.Mean(name="val_e_loss")
-            self.eval_metrics["val_e_f1"] = tf.keras.metrics.Mean(name="val_e_f1")
+
+            self.metric_f1_score = tfa.metrics.F1Score(
```
Comment (Ghostvv) [inline rasa/nlu/classifiers/embedding_intent_classifier.py]:
```suggestion
            self._crf_f1_score = tfa.metrics.F1Score(
```
we create for num_tags, so cannot use for anything else

File: rasa/utils/tensorflow/tf_models.py [hunk 3]
```diff
@@ -51,37 +48,89 @@ def fit(
...
+            tf_evaluation_dataset_function,
+            tf_evaluation_on_batch_function,
+        ) = self._get_tf_evaluation_functions(
+            eager, evaluate_on_num_examples, evaluation_model_data
+        )
+
+        pbar = tqdm(range(epochs), desc="Epochs", disable=disable)
+
+        for ep in pbar:
+            ep_batch_size = self.linearly_increasing_batch_size(ep, batch_size, epochs)
```
Comment (Ghostvv) [inline rasa/utils/tensorflow/tf_models.py]: needs to be a tensor, otherwise, `tf.function` magic wouldn't work

File: rasa/nlu/classifiers/embedding_intent_classifier.py [hunk 4]
```diff
@@ -909,16 +875,26 @@ def __init__(
...
+        self.mask_acc = tf.keras.metrics.Mean(name="m_acc")
+        self.entity_loss = tf.keras.metrics.Mean(name="e_loss")
+        self.entity_f1 = tf.keras.metrics.Mean(name="e_f1")
+
+        if self.config[INTENT_CLASSIFICATION]:
+            self.metrics_to_log += ["i_acc", "i_loss"]
+        if self.config[MASKED_LM]:
+            self.metrics_to_log += ["m_acc", "m_loss"]
+        if self.config[ENTITY_RECOGNITION]:
+            self.metrics_to_log += ["e_loss", "e_f1"]
```
Comment (Ghostvv) [inline rasa/nlu/classifiers/embedding_intent_classifier.py]: shall we do it inside corresponding `prepare...` methods?

File: rasa/nlu/classifiers/embedding_intent_classifier.py [hunk 5]
```diff
@@ -909,16 +875,26 @@ def __init__(
...
-        self.entity_f1 = tfa.metrics.F1Score(
-            num_classes=self._num_tags - 1,  # `0` prediction is not a prediction
-            average="micro",
-        )
+        self.intent_acc = tf.keras.metrics.Mean(name="i_acc")
+        self.intent_loss = tf.keras.metrics.Mean(name="i_loss")
+        self.mask_loss = tf.keras.metrics.Mean(name="m_loss")
+        self.mask_acc = tf.keras.metrics.Mean(name="m_acc")
+        self.entity_loss = tf.keras.metrics.Mean(name="e_loss")
+        self.entity_f1 = tf.keras.metrics.Mean(name="e_f1")
```
Comment (Ghostvv) [inline rasa/nlu/classifiers/embedding_intent_classifier.py]: so layers can be part of the dict but metrics not? Good job tensorflow, very intuitive 🙈

File: rasa/nlu/classifiers/embedding_intent_classifier.py [hunk 6]
```diff
@@ -966,11 +942,11 @@ def _prepare_layers(self) -> None:
         else:
             self._transformer = lambda x, mask, training: x
 
-        self._embed = {}
-
-        self.train_metrics = {"t_loss": tf.keras.metrics.Mean(name="t_loss")}
-        self.eval_metrics = {"val_t_loss": tf.keras.metrics.Mean(name="val_t_loss")}
+        self._prepare_mask_lm_layers()
+        self._prepare_intent_classification_layers()
+        self._prepare_entity_recognition_layers()
```
Comment (Ghostvv) [inline rasa/nlu/classifiers/embedding_intent_classifier.py]: I think it makes sense to create another method for sequence related staff above.
Shall we use a `dict` for all the layers including losses? Then we don't need `else`. Either general, or separate for each different method, not sure which approach is better. `self._layers` is occupied though.

File: rasa/nlu/classifiers/embedding_intent_classifier.py (hunk 5 above)
Comment (tabergma) [reply rasa/nlu/classifiers/embedding_intent_classifier.py]: yeah, if you want to collect the metrics afterwards, they need to be class variables...

File: rasa/nlu/classifiers/embedding_intent_classifier.py (hunk 2 above)
Comment (Ghostvv) [reply rasa/nlu/classifiers/embedding_intent_classifier.py]: we don't actually need to collect this metric, so it can be part of the dict

Comment (tabergma) [reply rasa/nlu/classifiers/embedding_intent_classifier.py]: there is no dict anymore, we can remove `TraningMetrics` once we adapted core

Comment (Ghostvv) [reply rasa/nlu/classifiers/embedding_intent_classifier.py]: by dict, I mean my suggestion about putting all the layers into a dict

Comment (tabergma) [reply rasa/nlu/classifiers/embedding_intent_classifier.py]: ah ok, will try that

File: rasa/nlu/classifiers/embedding_intent_classifier.py (hunk 6 above)
Comment (Ghostvv) [reply rasa/nlu/classifiers/embedding_intent_classifier.py]: and the metric that calculates f1 score for crf also can be inside this dict

Comment (tabergma) [reply rasa/nlu/classifiers/embedding_intent_classifier.py]: Are you sure that `self._layers` is occupied?

Comment (Ghostvv) [reply rasa/nlu/classifiers/embedding_intent_classifier.py]: oh, they might be occupied only for `tf.keras.layers.Layer`

Comment (tabergma) [reply rasa/nlu/classifiers/embedding_intent_classifier.py]: also used for `tf.keras.models.Model`

File: rasa/nlu/classifiers/embedding_intent_classifier.py [hunk 7]
```diff
@@ -963,84 +947,66 @@ def _prepare_layers(self) -> None:
...
-            self.train_metrics["m_loss"] = tf.keras.metrics.Mean(name="m_loss")
-            self.train_metrics["m_acc"] = tf.keras.metrics.Mean(name="m_acc")
-            self.eval_metrics["val_m_loss"] = tf.keras.metrics.Mean(name="val_m_loss")
-            self.eval_metrics["val_m_acc"] = tf.keras.metrics.Mean(name="val_m_acc")
-        else:
-            self._input_mask = None
-            self._loss_mask = None
+    def _prepare_mask_lm_layers(self):
+        self._tf_layers["input_mask"] = tf_layers.InputMask()
+        self._tf_layers["embed"]["text_mask"] = tf_layers.Embed(
```
Comment (Ghostvv) [inline rasa/nlu/classifiers/embedding_intent_classifier.py]:
```suggestion
        self._tf_layers["embed"]["lm_mask"] = tf_layers.Embed(
```
maybe it is more intuitive?

File: rasa/nlu/classifiers/embedding_intent_classifier.py [hunk 8]
```diff
@@ -963,84 +947,66 @@ def _prepare_layers(self) -> None:
...
+            self.config[EMBED_DIM],
+            self.config[C2],
+            "text_mask",
+            self.config[SIMILARITY_TYPE],
+        )
+        self._tf_layers["embed"]["text_token"] = tf_layers.Embed(
```
Comment (Ghostvv) [inline rasa/nlu/classifiers/embedding_intent_classifier.py]:
```suggestion
        self._tf_layers["embed"]["gold_token"] = tf_layers.Embed(
```
?

File: rasa/nlu/classifiers/embedding_intent_classifier.py [hunk 9]
```diff
@@ -909,19 +875,39 @@ def __init__(
...
+        if self.config[INTENT_CLASSIFICATION]:
+            self.metrics_to_log += ["i_loss", "i_acc"]
+        if self.config[ENTITY_RECOGNITION]:
+            self.metrics_to_log += ["e_loss", "e_f1"]
 
-        self._sparse_to_dense = {
+    def _prepare_layers(self) -> None:
+        self._tf_layers = {}
+        self._prepare_sequence_layers()
+        self._tf_layers["embed"] = {}
```
Comment (Ghostvv) [inline rasa/nlu/classifiers/embedding_intent_classifier.py]: do you think we need double dicts?

File: rasa/nlu/classifiers/embedding_intent_classifier.py [hunk 10]
```diff
@@ -963,84 +942,66 @@ def _prepare_layers(self) -> None:
...
-            self.eval_metrics["val_m_acc"] = tf.keras.metrics.Mean(name="val_m_acc")
-        else:
-            self._input_mask = None
-            self._loss_mask = None
+    def _prepare_mask_lm_layers(self) -> None:
+        self._tf_layers["input_mask"] = tf_layers.InputMask()
+        self._tf_layers["embed.lm_mask"] = tf_layers.Embed(
+            self.config[EMBED_DIM],
+            self.config[C2],
+            "text_mask",
```
Comment (Ghostvv) [inline rasa/nlu/classifiers/embedding_intent_classifier.py]:
```suggestion
            "lm_mask",
```
to keep the names consistent

File: rasa/nlu/classifiers/embedding_intent_classifier.py [hunk 11]
```diff
@@ -963,84 +942,66 @@ def _prepare_layers(self) -> None:
...
+            self.config[SIMILARITY_TYPE],
+        )
+        self._tf_layers["embed.golden_token"] = tf_layers.Embed(
+            self.config[EMBED_DIM],
+            self.config[C2],
+            "text_token",
```
Comment (Ghostvv) [inline rasa/nlu/classifiers/embedding_intent_classifier.py]:
```suggestion
            "golden_token",
```

File: rasa/nlu/classifiers/embedding_intent_classifier.py [hunk 12]
```diff
@@ -963,84 +942,66 @@ def _prepare_layers(self) -> None:
...
+            self.config[SIMILARITY_TYPE],
+        )
+        self._tf_layers["loss_mask"] = tf_layers.DotProductLoss(
```
Comment (Ghostvv) [inline rasa/nlu/classifiers/embedding_intent_classifier.py]:
```suggestion
        self._tf_layers["loss.mask"] = tf_layers.DotProductLoss(
```

File: rasa/nlu/classifiers/embedding_intent_classifier.py [hunk 13]
```diff
@@ -963,84 +942,66 @@ def _prepare_layers(self) -> None:
...
+            "text",
+            self.config[SIMILARITY_TYPE],
+        )
+        self._tf_layers["embed.label"] = tf_layers.Embed(
+            self.config[EMBED_DIM],
+            self.config[C2],
+            "label",
+            self.config[SIMILARITY_TYPE],
+        )
+        self._tf_layers["loss_label"] = tf_layers.DotProductLoss(
```
Comment (Ghostvv) [inline rasa/nlu/classifiers/embedding_intent_classifier.py]:
```suggestion
        self._tf_layers["loss.label"] = tf_layers.DotProductLoss(
```

File: rasa/nlu/classifiers/embedding_intent_classifier.py [hunk 14]
```diff
@@ -163,14 +163,14 @@ class EmbeddingIntentClassifier(EntityExtractor):
         # how often to calculate training accuracy
         EVAL_NUM_EPOCHS: 20,  # small values may hurt performance
         # how many examples to use for calculation of training accuracy
-        EVAL_NUM_EXAMPLES: 0,  # large values may hurt performance
+        EVAL_NUM_EXAMPLES: 10,  # large values may hurt performance
```
Comment (Ghostvv) [inline rasa/nlu/classifiers/embedding_intent_classifier.py]: let's keep default to `0`, since it excludes some data from training

Comment (tabergma) [reply rasa/nlu/classifiers/embedding_intent_classifier.py]: reverted the change in the last commit
